src/helper/sampler.hpp

```diff
@@ -0,0 +1,91 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog
+// (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+
+#pragma once
+
+#include <atomic>
+#include <cmath>
+#include <cstdint>
+#include <iostream>
+#include <mutex>
+#include <optional>
+
+namespace dds {
+static const double min_rate = 0.0001;
+class sampler {
+public:
+    sampler(double sample_rate) : sample_rate_(sample_rate)
+    {
+        // NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        if (sample_rate_ <= 0) {
+            sample_rate_ = 0;
+        } else if (sample_rate_ > 1) {
+            sample_rate_ = 1;
+        } else if (sample_rate_ < min_rate) {
+            sample_rate_ = min_rate;
+        }
+        // NOLINTEND(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+    }
+    class scope {
+    public:
+        explicit scope(std::atomic<bool> &concurrent) : concurrent_(&concurrent)
+        {
+            *concurrent_ = true;
```

Comment:
This could probably be done with relaxed memory order since this is typically only created within a mutex which already provides synchronisation:

```
concurrent_.store(true, std::memory_order_relaxed)